ragnarok_formats/src/map.rs

```diff
@@ -55,6 +72,79 @@ bitflags::bitflags! {
+        const MAX_DEPTH: usize = 5;
+        const CHILD_COUNT: usize = 4;
+
+        // You can check in the Map Viewer in Debugging
+        // Simulate a DFS using a stack.
+        let mut stack = VecDeque::<(usize, usize)>::from([(0, 0)]);
+        let mut nodes = HashMap::new();
+        let mut node_parents: Vec<usize> = Vec::new();
+
+        while let Some((parent_node, depth)) = stack.pop_back() {
+            // Read the input
+            let max: [f32; 3] = FromBytes::from_bytes(byte_stream).trace::<Self>()?;
+            let min: [f32; 3] = FromBytes::from_bytes(byte_stream).trace::<Self>()?;
+            let half_size: [f32; 3] = FromBytes::from_bytes(byte_stream).trace::<Self>()?;
+            let center: [f32; 3] = FromBytes::from_bytes(byte_stream).trace::<Self>()?;
+
+            // Allocate the vector to get a limited amount of child and crash if there is
+            // more children.
+            let children: Vec<QuadTreeData> = match depth < MAX_DEPTH {
+                true => Vec::with_capacity(CHILD_COUNT),
+                false => Vec::new(),
+            };
+
+            // Put the position of index parent_node in the set
+            nodes.insert(parent_node, QuadTreeData {
+                max,
+                min,
+                half_size,
+                center,
+                children,
+            });
+
+            if depth < MAX_DEPTH {
+                let self_index = node_parents.len();
+                // First add the nodes and the corresponding parents
+                for _counter in 0..CHILD_COUNT {
+                    node_parents.push(parent_node);
+                }
+                // Observe that we are inserting the position in reverse order.
+                // This is because the first element that we want to insert to the stack is the
+                // last to enter the stack.
+                for counter in 0..CHILD_COUNT {
+                    let revert_position = CHILD_COUNT - counter;
+                    stack.push_back((self_index + revert_position, depth + 1));
+                }
+            }
+        }
+
+        // Iterate nodes backwards and assign them the their parents.
+        let mut child_index = node_parents.len();
+        for parent_index in node_parents.into_iter().rev() {
+            // Get the children by their index and corresponding parent
+            let mut child = nodes.remove(&child_index).unwrap();
+            child.children.reverse();
+            child_index -= 1;
+            let mut parent = nodes.remove(&parent_index).unwrap();
+            parent.children.push(child);
+            nodes.insert(parent_index, parent);
+        }
+
+        // The root haven't reverse yet, so we need to reverse it
+        // Observe that the root isn't in parent_index, only the childs of root.
+        assert_eq!(nodes.len(), 1);
+        let mut root: QuadTreeData = nodes.remove(&0).unwrap();
+        root.children.reverse();
+
+        Ok(root)
```

Comment:
I did some refactoring on the newer version again, please check it out:

```rust
        // Helper struct to keep node accesses readable.
        struct Node {
            data: QuadTreeData,
            parent_index: usize,
        }

        // Simulate a DFS using a stack.
        let mut stack = VecDeque::from([(0, 0)]);
        let mut nodes = Vec::new();

        while let Some((parent_index, depth)) = stack.pop_back() {
            let max = FromBytes::from_bytes(byte_stream).trace::<Self>()?;
            let min = FromBytes::from_bytes(byte_stream).trace::<Self>()?;
            let half_size = FromBytes::from_bytes(byte_stream).trace::<Self>()?;
            let center = FromBytes::from_bytes(byte_stream).trace::<Self>()?;

            let children = match depth < MAX_DEPTH {
                true => Vec::with_capacity(CHILD_COUNT),
                false => Vec::new(),
            };

            // Add next node lookups to the stack.
            if depth < MAX_DEPTH {
                let self_index = nodes.len();

                for _counter in 0..CHILD_COUNT {
                    stack.push_back((self_index, depth + 1));
                }
            }

            let data = QuadTreeData {
                max,
                min,
                half_size,
                center,
                children,
            };
            nodes.push(Node { data, parent_index });
        }

        // Go back to front and assign nodes to their parents. Node 0 is the root
        // node so we skip that.
        while nodes.len() > 1 {
            let node = nodes.pop().unwrap();
            nodes[node.parent_index].data.children.insert(0, node.data);
        }

        let root_node = nodes.pop().unwrap().data;
        Ok(root_node)
```

Notable differences:
- Went back to using a `Vec` for the nodes since `HashMap` is not a good data structure for this use case
- Use the same storage for the nodes and the parent indices since they have a 1 to 1 mapping (introduced a helper struct `Node` to help with keeping this structured)
- Using `insert(0)` instead of `push()` followed by `reverse`
- Let Rust infer the types (`stack`, `min`, `max`, `children`, etc.). Rust has very good type inference, let it do the work!

I have confirmed through testing that this current version creates the exact same quad tree as your code.

Reply:
Oh, okay, it is much more neat than I expected. With the modification that you made, the code becomes more neat.
I need to process if everything is alright, because the nodes vector, may be not using the correct index. Need to
confirm by hand with the program running. Later, I will return if the code is working and commit the code.

> Let Rust infer the types (`stack`, `min`, `max`, `children`, etc.). Rust has very good type inference, let it do the work!
I was trying to print the elements and was not detecting the type after `FromBytes::from_bytes(byte_stream).trace::<Self>()?;` doesn't know the reason, but I will remember to remove in the next commit.

> Using insert(0) instead of push() followed by reverse
Strategy 1. Using this method of insert to begin
When you insert to begin, you shift all elements from vector by 1 and place the element. 
For insertion, you use 4 operation. For the shift all elements from vector by, you would use (1+2+3) operation 6 operation. Resulting in 10 operation
Strategy 2. Using this method of insert to end and reverse, you only shift during reverse.
For insertion, you use 4 operation. And for reverse, you use more 4 operation. Resulting in 8 operations.

For code understanding, the strategy 1 will be easier to understand.